### PR TITLE
[autoscaler] Additional Autoscaler Metrics

### PR DESCRIPTION
**Pull request by ckw017**
## Why are these changes needed?

Followup to #16066. Adds the following metrics
- `worker_launch_time` (histogram), measured as time for provider.create_node to execute
- `worker_update_time` (histogram), measured as time for NodeLauncherThread.run() to successfully execute
- `updating_nodes` (gauge), length of StandardAutoscaler.updaters
- `failed_create_nodes` (counter), incremented whenever by number of nodes in batch whenever provider.create_node throws an exception
- `failed_updates` (counter), tracks StandardAutoscaler.num_failed_updates
- `successful_updates` (counter) tracks StandardAutoscaler.num_successful_updates

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/_private/node_launcher.py [hunk 1]
```diff
@@ -74,6 +83,7 @@ def run(self):
                 self._launch_node(config, count, node_type)
             except Exception:
                 self.prom_metrics.node_launch_exceptions.inc()
+                self.prom_metrics.failed_create_nodes.inc(count)
```
Comment (ijrsvt) [inline python/ray/autoscaler/_private/node_launcher.py:86]: Are these metrics the same?

File: python/ray/autoscaler/_private/autoscaler.py [hunk 2]
```diff
@@ -271,13 +271,17 @@ def _update(self):
             for node_id in completed_nodes:
                 if self.updaters[node_id].exitcode == 0:
                     self.num_successful_updates[node_id] += 1
+                    self.prom_metrics.successful_updates.inc()
+                    update_time = self.updaters[node_id].update_time
```
Comment (ijrsvt) [inline python/ray/autoscaler/_private/autoscaler.py:275]: Can we do a check that `update_time` is not None?

**Review by ijrsvt — COMMENTED** [review]
two quick comments!

File: python/ray/autoscaler/_private/node_launcher.py (hunk 1 above)
Comment (ckw017) [reply python/ray/autoscaler/_private/node_launcher.py:86]: Sort of, failed_create_nodes increments by the number of nodes in the batch whenever a launch fails (number of nodes we tried to launch, but failed due to exception), while node_launch_exceptions always increments by 1 (number of exceptions). So if we only launch 1 node at a time they would be the same

Comment (ijrsvt) [reply python/ray/autoscaler/_private/node_launcher.py:86]: OOps, totally missed this!

**Review by ijrsvt — APPROVED** [review]
LGTM!

**Comment by ckw017** [comment]
@ijrsvt ah sorry, I still have the problem where I don't have auth to merge my own PRs. If you're good with this you can merge when you can go forward with the merge, otherwise we can wait for more reviewers

**Comment by ijrsvt** [comment]
@ckw017 Let's wait and see if @AmeerHajAli wants to say anythin!

File: python/ray/autoscaler/_private/autoscaler.py [hunk 3]
```diff
@@ -332,6 +338,7 @@ def _update(self):
         for node_id in nodes:
```
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/autoscaler.py:338]: Should we add here a tracking of recovering nodes?

File: python/ray/autoscaler/_private/node_launcher.py [hunk 4]
```diff
@@ -60,7 +61,15 @@ def _launch_node(self, config: Dict[str, Any], count: int,
...
             node_config.update(launch_config)
+        launch_start_time = time.time()
         self.provider.create_node(node_config, node_tags, count)
+        launch_time = time.time() - launch_start_time
+        for _ in range(count):
+            # Note: when launching multiple nodes we observe the time it
+            # took all nodes to launch for each node. For example, if 4
+            # nodes were launched in 25 seconds, we would observe the 25
+            # second launch time 4 times.
+            self.prom_metrics.worker_launch_time.observe(launch_time)
```
Comment (AmeerHajAli) [inline python/ray/autoscaler/_private/node_launcher.py:72]:
```suggestion
            self.prom_metrics.worker_create_node_time.observe(launch_time)
```

**Review by AmeerHajAli — COMMENTED** [review]
LGTM overall. Left a few minor comments.

**Pull request merged**